pageserver/src/walreceiver/walreceiver_connection.rs

```diff
@@ -154,7 +154,7 @@ pub async fn handle_walreceiver_connection(
 
                 {
                     let mut decoded = DecodedWALRecord::default();
-                    let mut modification = timeline.begin_modification();
+                    let mut modification = timeline.begin_modification(endlsn);
```

Comment:
> I am not sure if saving 8 bytes per modification is enough motivation for not storing lsn and making implementation of relation size cache in LayeredTimeline problematic...

To provide context, it was not about saving 8 bytes, but more about reusing `DatadirModification`.  In the below code, the `modification` is reused for each walingest. Noticed that the `lsn` is different while in the loop.

In other words, this change is not enough. You should at least update `modification.lsn = lsn` in the loop.